File: Blind 75/11 - Graphs/323-number-of-connected-components-in-an-undirected-graph.py
'''
*LeetCode premium problem
'''
from time import time
from typing import List


class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        node_map = {i: [] for i in range(n)}
        for a, b in edges:
            node_map[a].append(b)

        visited = set()

        def dfs(node):
            if node not in visited:
                visited.add(node)
                for neighbor in node_map[node]:
                    dfs(neighbor)

            return 1

        def bfs(q):
            for node in q:
                if node not in visited:
                    q += node_map[node]
                    visited.add(node)

            return 1

        return sum(dfs(i) for i in range(n) if i not in visited)
        return sum(bfs([i]) for i in range(n) if i not in visited)

    def countComponentsUnionFind(self, n: int, edges: List[List[int]]) -> int:
        p = range(n)

        def find(v):
            if p[v] != v:
                p[v] = find(p[v])

            return p[v]

        for v, w in edges:
            p[find(v)] = find(w)

        return len(set(map(find, p)))

    def reference(self, n: int, edges: List[List[int]]) -> int:
        parent_arr = [i for i in range(n)]
        rank = [1] * n

        def find(n1):
            res = n1
            while res != parent_arr[res]:
                parent_arr[res] = parent_arr[parent_arr[res]]
                res = parent_arr[res]
            return res

        def union(n1, n2):
            p1, p2 = find(n1), find(n2)

            if p1 == p2:
                return 0

            if rank[p2] > rank[p1]:
                parent_arr[p1] = p2
                rank[p2] += rank[p1]
            else:
                parent_arr[p2] = p1
                rank[p1] += rank[p2]
            return 1

        res = n
        for n1, n2 in edges:
            res -= union(n1, n2)
        return res

    def quantify(self, test_cases, runs=100000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.countComponents(case[0], case[1]))
                else:
                    self.countComponents(case[0], case[1])
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case[0], case[1]))
                else:
                    self.reference(case[0], case[1])
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        (5,
         [
             [0, 1],
             [1, 2],
             [3, 4]
         ])
    ]
    test.quantify(test_cases)

File: Blind 75/11 - Graphs/JavaScript/323-number-of-connected-components-in-an-undirected-graph.js
const { performance } = require('perf_hooks');

class Solution {
  method(n, edges) {
    let adjList = Array.from({ length: n }, () => []);
    for (const [a, b] of edges) {
      adjList[a].push(b);
    }

    let seen = new Set();

    const dfs = (node) => {
      if (seen.has(node)) return false;
      seen.add(node);
      for (const n of adjList[node]) {
        dfs(n);
      }

      return true;
    };

    let count = 0;
    for (const [k, v] of adjList.entries()) {
      if (dfs(k)) count += 1;
    }

    return count;
  }

  reference(n, edges, count = 0) {
    const { graph, visited } = this.buildGraph(n, edges);

    for (const node in graph) {
      if (this.hasPath(graph, node, visited)) count++;
    }

    return count;
  }

  initGraph = (n) => ({
    graph: new Array(n).fill().map(() => []),
    visited: new Array(n).fill(false),
  });

  buildGraph = (n, edges) => {
    const { graph, visited } = this.initGraph(n);

    for (const [src, dst] of edges) {
      graph[src].push(dst);
      graph[dst].push(src);
    }

    return { graph, visited };
  };

  hasPath = (graph, current, visited) => {
    if (visited[current]) return false;
    visited[current] = current;

    for (const neighbor of graph[current]) {
      this.hasPath(graph, neighbor, visited);
    }

    return true;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    5,
    [
      [0, 1],
      [1, 2],
      [3, 4],
    ],
  ],
];
test.quantify(testCases);
